orchagent/nvgreorch.cpp

```diff
@@ -0,0 +1,515 @@
+    attr.value.objlist.count = num_encap_map;
+    attr.value.objlist.list = encap_map_list;
+    tunnel_attrs.push_back(attr);
+
+    attr.id = SAI_TUNNEL_ATTR_ENCAP_SRC_IP;
+    attr.value.ipaddr = src_ip;
+    tunnel_attrs.push_back(attr);
+
+    sai_object_id_t tunnel_id;
+    sai_status_t status = sai_tunnel_api->create_tunnel(
+                                &tunnel_id,
+                                gSwitchId,
+                                static_cast<uint32_t>(tunnel_attrs.size()),
+                                tunnel_attrs.data()
+                          );
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        throw std::runtime_error("Can't create the NVGRE tunnel object");
+    }
+
+    return tunnel_id;
+}
+
+/** @brief Removes tunnel in SAI.
+ *
+ *  @param tunnel_id Pointer to tunnel identifier.
+ *
+ *  @return void.
+ */
+void NvgreTunnel::sai_remove_tunnel(sai_object_id_t tunnel_id)
+{
+    sai_status_t status = sai_tunnel_api->remove_tunnel(tunnel_id);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        throw std::runtime_error("Can't remove the NVGRE tunnel object");
+    }
+}
+
+void NvgreTunnel::createNvgreMappers()
+{
+    for (auto map_type : nvgreMapTypes)
+    {
+        tunnel_ids_.tunnel_encap_id.insert(
+            make_pair(map_type, sai_create_tunnel_map(get_encap_nvgre_mapper(map_type)))
+        );
+    }
+
+    for (auto map_type : nvgreMapTypes)
+    {
+        tunnel_ids_.tunnel_decap_id.insert(
+            make_pair(map_type, sai_create_tunnel_map(get_decap_nvgre_mapper(map_type)))
+        );
+    }
+}
+
+void NvgreTunnel::removeNvgreMappers()
+{
+    for (auto map_type : nvgreMapTypes)
+    {
+        sai_remove_tunnel_map(getEncapMapId(map_type));
+    }
+
+    for (auto map_type : nvgreMapTypes)
+    {
+        sai_remove_tunnel_map(getDecapMapId(map_type));
+    }
+
+    tunnel_ids_.tunnel_encap_id.clear();
+    tunnel_ids_.tunnel_decap_id.clear();
+}
+
+void NvgreTunnel::createNvgreTunnel()
+{
+    sai_ip_address_t ip_addr;
+    swss::copy(ip_addr, src_ip_);
+
+    tunnel_ids_.tunnel_id = sai_create_tunnel(tunnel_ids_, ip_addr, gUnderlayIfId);
+
+    SWSS_LOG_INFO("NVGRE tunnel '%s' was created", tunnel_name_.c_str());
+}
+
+void NvgreTunnel::removeNvgreTunnel()
+{
+    try
+    {
+        sai_remove_tunnel(tunnel_ids_.tunnel_id);
+    }
+    catch(const std::runtime_error& error)
+    {
+        SWSS_LOG_ERROR("Error while removing tunnel entry. Tunnel: %s. Error: %s", tunnel_name_.c_str(), error.what());
+    }
+
+    SWSS_LOG_INFO("NVGRE tunnel '%s' was removed", tunnel_name_.c_str());
+
+    tunnel_ids_.tunnel_id = SAI_NULL_OBJECT_ID;
+}
+
+NvgreTunnel::NvgreTunnel(std::string tunnelName, IpAddress srcIp) :
+                         tunnel_name_(tunnelName),
+                         src_ip_(srcIp)
+{
+    createNvgreMappers();
+    createNvgreTunnel();
+}
+
+NvgreTunnel::~NvgreTunnel()
+{
+    removeNvgreTunnel();
+    removeNvgreMappers();
+}
+
```

Comment:
nit:
Why does the code intersperse NvgreTunnel and NvgreTunnelOrch code?
It would be better to complete NvgreTunnel code first before moving on to NvgreTunnelOrch

Reply:
I took `VxLAN` orchagent implementation as an example.